admin-tools/man/euctl.1
.Dd 2017-01-12
.Dt EUCTL 1
.Os eucalyptus 4.4
.Sh NAME
.Nm euctl
.Nd get or set cloud state
.Sh SYNOPSIS
.Nm
.Op Fl A\&nr
.Op Fl d | s
.Ar NAME ...
.Nm
.Op Fl nq
.Ar NAME=VALUE ...
.Nm
.Op Fl nq
.Ar NAME=@FILE ...
.Nm
.Fl -dump
.Op Fl -format Cm {raw,json,yaml}
.Ar NAME
.Nm
.Fl -edit
.Op Fl -format Cm {raw,json,yaml}
.Ar NAME
.Sh DESCRIPTION
The
.Nm
utility retrieves cloud state and allows cloud administrators
to set cloud state.  The state to be retrieved or set is
described using a
.Dq Management Information Base
.Pq MIB
style name, described as a dotted set of components.
.Pp
When retrieving a variable, a subset of the MIB name may be
specified to retrieve a list of variables in that subset.
For example, to list all the dns variables:
.Pp
.Dl $ Nm No dns
.Pp
When setting a variable, the MIB name should be followed by
an equal sign and the new value:
.Pp
.Dl $ Nm No dns.enabled=true
.Pp
To reset a varable to its default value, specify the
.Fl r
option:
.Pp
.Dl $ Nm Fl r No dns.enabled
.Pp
The information available from euctl consists of integers,
strings, and structures.  Structures can only be retrieved
by specialized programs and, in some cases, this program's
.Fl -edit
and
.Fl -dump
options.
.Ss Positional arguments:
.Bl -tag -width Ds
.It Ar NAME
Output a variable's value.
.It Ar NAME=VALUE
Set a variable to the specified value and then output it.
.It Ar NAME=@FILE
Set a variable to that of the specified file's contents,
then output it.
.El
.Ss Optional arguments:
.Bl -tag -width Ds
.It Fl A , Fl -all-types
List all the known variable names, including
structures.  Those with string or integer values will
be output as usual; for the structures, the methods of
retrieving them are given.
.It Fl d
Output variables' default values rather than their
current values.  Note that not all variables have default
values.
.It Fl s
Output variables' descriptions rather than their current
values.
.It Fl n
Suppress output of variable names. This is useful
for setting shell variables.
.It Fl q
Suppress all output when setting a variable. This
option overrides the behavior of
.Fl n .
.It Fl r , Fl -reset
Reset the given variables to their default values.
.It Fl -dump
Output the value of a structure variable in its
entirety.  The value will be formatted in the manner
specified by the
.Fl -format
option.
.It Fl -edit
Edit the value of a structure variable interactively.
The value will be formatted in the manner specified by the
.Fl -format
option.  Only one variable may be edited per
invocation.  When looking for an editor, the program
will first try the environment variable
.Ev VISUAL ,
then the environment variable
.Ev EDITOR ,
and finally the
default editor,
.Xr vi 1 .
.It Fl -format Cm {raw,json,yaml}
Use the specified format when displaying a structure.
(default:
.Cm json Ns )
.It Fl h , Fl -help
Show a help message and exit without doing anything.
.It Fl U Ar URL , Fl -url Ar URL
The URL of the service to contact.
.It Fl I Ar KEY_ID , Fl -access-key-id Ar KEY_ID
The access key ID to use to authenticate requests.
.It Fl S Ar KEY . Fl -secret-key Ar KEY
The secret key to use to authenticate requests.
.It Fl -security-token Ar TOKEN
The STS security token to use to authenticate requests.
.It Fl -region Ar USER@REGION
The user and/or region name to use when looking up
configuration file data.
.It Fl -debug
Show debugging output when running.
.It Fl -debugger
Launch an interactive epdb or pdb debugger after
encountering
an error.
.It Fl -version
show the program's version and exit
.El
.Sh ENVIRONMENT
.Bl -tag -width Ds
.It Ev AWS_ACCESS_KEY_ID
The access key ID to use when authenticating web
service requests.  This takes precedence over
.Fl -region
and
.Xr euca2ools.ini 5 ,
but not
.Fl I .
.It Ev AWS_SECRET_ACCESS_KEY
The secret key to use when authenticating web
service requests.  This takes precedence over
.Fl -region
and
.Xr euca2ools.ini 5 ,
but not
.Fl S .
.It Ev EUCA_PROPERTIES_URL
The URL of the service to contact.  This takes precedence
over
.Fl -region
and
.Xr euca2ools.ini 5 ,
but not
.Fl U .
.It Ev VISUAL
The editor to invoke when using the
.Fl -edit
option.
.It Ev EDITOR
The editor to invoke when using the
.Fl -edit
option when
.Ev VISUAL is not set.
.El
.Sh SEE ALSO
.Xr euca2ools.ini 5 ,
.Xr eucalyptus-admin 7
.Sh HISTORY
The
.Nm
command first appeared in Eucalyptus 4.2.
.Sh NOTES
.Nm
requires access keys and knowledge of where to locate
the web services it needs to contact.  It can obtain
these from several locations.  See
.Xr eucalyptus-admin 7
for more information.


CHANGELOG
*Wed Aug 28 2012 Eucalyptus Systems (support@eucalyptus.com)
    * Version 3.1.1
    * Resolves the following issues:
        EUCA-280  CLC assigns a public IP to VMs after elastic IP association/disassociation while using --addressing=private
        EUCA-304  Ability to customize "eucalyptus" string in DNS population
        EUCA-1473 DescribeInstancesResponse timestamps have the wrong format
        EUCA-1492 Delete on terminate false is always reported as true
        EUCA-1636 DescribeSnapshotsResponse timestamps have the wrong format
        EUCA-1792 CLC accepts RunInstances requests with MinCount > MaxCount
        EUCA-2002 Allocating and assigning addresses doesn't work
        EUCA-2024 Special character in Filename and inconsistent behavior of Walrus
        EUCA-2046 Policy using "numlteq" is rejected (short form of NumericLessThanEquals)
        EUCA-2204 euca-disassociate-address ends up giving new public ip address to instance
        EUCA-2260 postgresql-binaries.properties.in should not be in eucalyptus-postgresql.jar
        EUCA-2311 Databases are created with LATIN1 encoding
        EUCA-2327 Duplicate address assignments to the same instance
        EUCA-2333 EIAM Extension Quota Policy Issue => ec2:quota-vminstancenumber
        EUCA-2340 VMware Broker: XML config option "vswitch" not honored
        EUCA-2345 cloud.vmstate.instance_subdomain is not honored by CLC
        EUCA-2413 NC Crashes when bridge device isn't configured
        EUCA-2417 issues upgrading to Eucalyptus 3.1.0 on the server running CLC, Walrus, CC, SC
        EUCA-2418 upgrade from 3.0.2 to 3.1 fails complains quota policy has got NumericLessThanEquals
        EUCA-2430 invalid error message when run instance exceed quota
        EUCA-2445 First Time Login via WebUI - No Check Old Password/New Password
        EUCA-2547 HMAC signature verification fails for empty or absent parameter value
        EUCA-3066 Upgrade fails if Netapp snapshot directory exists
        EUCA-3270 CentOS 6: Cleanstop on CC not executing cleanup procedure
        EUCA-3313 Ability to customize "eucalyptus" string in DNS population
        EUCA-3314 Special character in Filename and inconsistent behavior of Walrus
        EUCA-3315 Allocating and assigning addresses doesn't work

*Wed Jul 7 2012 Eucalyptus Systems (support@eucalyptus.com)
    * Version 3.1.0
    * New Features:
        - New packaged platform support (Added support for CentOS 6.x
          and Ubuntu 12.04 LTS)
        - vSphere 5 support (plugin only).
    * Resolves the following issues:
        EUCA-587  Exclude .snapshot directories during upgrade process
        EUCA-808  clc: add back support for '--addressing private'
        EUCA-1070 Walrus: Overwritting object causes object to become inaccessible
        EUCA-1383 Issue when reconfiguring cc vnet ip range and mask
        EUCA-1487 eee: euca-describe-property occasionally exits with an OOM
        EUCA-1497 cc: centos kernel oops if too many vlan interfaces left enabled
        EUCA-1500 deleteOnTerminate=true doesn't delete a volume
        EUCA-1513 use of wildcard chars in params to services/Eucalyptus URL allows to detect valid AWSAccessKeyId
        EUCA-1514 MySQL connections are refused in some cases
        EUCA-1515 initial db sync retry logic can hang
        EUCA-1528 review all uses of db queries with LIKE condition
        EUCA-1546 CC <-> NC comm fails when number of volume records on an instance exceeds 64
        EUCA-1554 IAM: Account admin can not see the policies of users and groups within account
        EUCA-1563 euare: uploading new policy with the same name does not overwrite old policy
        EUCA-1596 Upgrade needs to migrate images after all users have been migrated
        EUCA-1621 StartInstance fails for SYSTEM/STATIC mode
        EUCA-1629 if instances are terminated quickly after being launched and then new instances are launched right away, new instances do not get public IPs
        EUCA-1647 Upgrade: After upgrade on RHEL5, euca-describe-instances shows public key instead of keypair name
        EUCA-1689 CC: metadata redirect is not re-applied in non HA mode if CC is DISABLED/ENABLED and pubinterface == privinterface
        EUCA-1693 EUARE: a salt should be used when hashing and storing user passwords
        EUCA-1694 NC: under load, NC fails to respond to Describe* requests promptly
        EUCA-1700 eustore-install-image should not download an image into a predictable filename in /tmp
        EUCA-1701 eustore-install-image should not untar and bundle images using predictable dir and filenames in /tmp
        EUCA-1710 cluster-pk.pem should not be on NC host
        EUCA-1713 SSL/TLS services should support cipher configuration and disable weak ciphers by default
        EUCA-1714 Secret key generation can have insufficient random data
        EUCA-1717 Walrus: anyone can access objects on Walrus by submitting correctly signed requests
        EUCA-1723 VMWareBroker: anyone can send SOAP messages to VMWareBroker
        EUCA-1735 CLC Reports incorrect root-device-type
        EUCA-1738 Walrus: DeleteWalrusSnapshot is missing an authorization check
        EUCA-1749 Non-SOAP requests can be replayed by changing the Signature "format"
        EUCA-1759 Let eucalyptus build when given the jars that it needs, not a file called cloud-lib.tar.gz
        EUCA-1766 HA CC upgrade: Disabled CC continues to have NAT rules for an instance

admin-tools/eucalyptus_admin/commands/properties/euctl.py
# Copyright (c) 2015-2017 Hewlett Packard Enterprise Development LP
#
# Redistribution and use of this software in source and binary forms,
# with or without modification, are permitted provided that the following
# conditions are met:
#
#   Redistributions of source code must retain the above copyright notice,
#   this list of conditions and the following disclaimer.
#
#   Redistributions in binary form must reproduce the above copyright
#   notice, this list of conditions and the following disclaimer in the
#   documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import json
import os
import subprocess
import sys
import tempfile
import textwrap

from requestbuilder import Arg, EMPTY, MutuallyExclusiveArgList
from requestbuilder.exceptions import ArgumentError
import yaml

from eucalyptus_admin.commands.properties import PropertiesRequest
from eucalyptus_admin.commands.properties.describeproperties import \
    DescribeProperties
from eucalyptus_admin.commands.properties.modifypropertyvalue import \
    ModifyPropertyValue


PROPERTY_TYPES = {'authentication.ldap_integration_configuration': 'json',
                  'cloud.network.network_configuration': 'json',
                  'cloudformation.swf_client_config': 'json',
                  'cloudformation.swf_workflow_worker_config': 'json',
                  'region.region_configuration': 'json'}


def _property_key_value(kvp_str):
    if '=' in kvp_str:
        key, val = kvp_str.split('=', 1)
        if not key:
            return (None, None)
        if val.startswith('@'):
            with open(val[1:]) as val_file:
                val = val_file.read()
        return (key.strip(), val.strip())
    else:
        return (kvp_str.strip(), None)


class Euctl(PropertiesRequest):
    DESCRIPTION = textwrap.dedent(
        '''\
        The %(prog)s utility retrieves cloud state and allows cloud
        administrators to set cloud state.  The state to be retrieved
        or set is described using a ``Management Information Base''
        (MIB) style name, described as a dotted set of components.

        When retrieving a variable, a subset of the MIB name may be
        specified to retrieve a list of variables in that subset.  For
        example, to list all the dns variables:

            $ %(prog)s dns

        When setting a variable, the MIB name should be followed by an
        equal sign (=) and the new value:

            $ %(prog)s dns.enabled=true

        To reset a varable to its default value, specify the -r option:

            $ %(prog)s -r dns.enabled

        The information available from %(prog)s consists of integers,
        strings, and structures.  The structures can only be retrieved
        by specialized programs and, in some cases, this program's
        --edit and --dump options.
        ''')

    ARGS = [Arg('prop_pairs', metavar='NAME[=VALUE|=@FILE]', nargs='*',
                type=_property_key_value,
                help='''Output the specified variable, and where a value
                is given, attempt to set it to the specified value.  If
                the value begins with an @ character then instead use
                the contents of the specified file as the variable's
                new value.'''),
            Arg('-A', '--all-types', action='store_true', help='''List all
                the known variable names, including structures.  Those with
                string or integer values will be output as usual; for the
                structures, the methods of retrieving them are given.'''),
            Arg('-r', '--reset', action='store_true',
                help='Reset the given variables to their default values.'),
            MutuallyExclusiveArgList(
                Arg('-d', dest='show_defaults', action='store_true',
                    help='''Show variables' default values instead of
                    their current values.'''),
                Arg('-s', dest='show_descriptions',
                    action='store_true', help='''Show variables'
                    descriptions instead of their current values.''')),
            Arg('-n', dest='suppress_name', action='store_true',
                help='''Suppress output of the variable name.  This is
                useful for setting shell variables.'''),
            Arg('-q', dest='suppress_all', action='store_true',
                help='''Suppress all output when setting a variable.  This
                option overrides the behavior of -n.'''),
            MutuallyExclusiveArgList(
                Arg('--edit', action='store_true', help='''Edit a structure
                    variable interactively.  Only one variable may be edited
                    per invocation.  When looking for an editor, the program
                    will first try the environment variable VISUAL, then the
                    environment variable EDITOR, and finally the default
                    editor, vi(1).'''),
                Arg('--dump', action='store_true', help='''Output the value of
                    a structure variable in its entirety.''')),
            Arg('--format', choices=('json', 'yaml', 'raw'), default='json',
                help='''Try to use the specified format when displaying
                a structure variable.  "raw" is only allowed with --dump.
                (default: json)''')]

    def configure(self):
        PropertiesRequest.configure(self)
        argcheck = None
        if self.args.get('dump'):
            argcheck = '--dump'
        elif self.args.get('edit'):
            argcheck = '--edit'
            if self.args.get('format') == 'raw':
                raise ArgumentError(
                    'argument --format: "raw" is only allowed with --dump')
        if argcheck:
            if len(self.args.get('prop_pairs') or []) != 1:
                raise ArgumentError(
                    'argument {0} must be accompanied by exactly one '
                    'variable name'.format(argcheck))
            if self.args.get('prop_pairs')[0][1]:
                raise ArgumentError(
                    'argument {0}: a value may not accompany the '
                    'variable name'.format(argcheck))
            if self.args.get('prop_pairs')[0][0] not in PROPERTY_TYPES:
                raise ArgumentError(
                    'argument {0} may not be used with variable {1}'
                    .format(argcheck, self.args['prop_pairs'][0][0]))

    def main(self):
        # FIXME:  This doesn't handle empty values.
        if self.args.get('dump'):
            prop_name = self.args.get('prop_pairs')[0][0]
            self.log.info('dumping property value   %s', prop_name)
            if self.args.get('format') == 'raw':
                # We need to actively prevent this from being parsed.
                prop_type = _Property
            else:
                prop_type = None
            prop = self._get_unique_property(prop_name, prop_type=prop_type)
            formatter = self._get_formatter(prop)
            print formatter.dumps(prop.value).strip()
        elif self.args.get('edit'):
            prop_name = self.args.get('prop_pairs')[0][0]
            self.log.info('editing property value   %s', prop_name)
            prop = self._get_unique_property(prop_name)
            has_changed = self.edit_interactively(prop)
            if has_changed:
                self.log.info('setting property value   %s', prop_name)
                self._set_property(prop_name, value=prop.dumps())
            else:
                print >> sys.stderr, prop_name, 'is unchanged'
        else:
            for key, val in self.args.get('prop_pairs') or [(None, None)]:
                if key == 'euca':
                    # DescribeProperties doesn't work on the magical,
                    # groovy code-injecting "euca" property, so we have
                    # to handle input and output in one shot.
                    #
                    # We don't do this all the time is case the value
                    # we chose doesn't actually get set for some reason.
                    self.log.info('handling one-shot property "euca"')
                    req = ModifyPropertyValue.from_other(self, Name=key,
                                                         Value=val)
                    response = req.main()
                    print response.get('value')
                    continue
                if val is not None:
                    self.log.info('setting property value   %s: %s', key, val)
                    self._set_property(key, value=val)
                if val is None and self.args.get('reset'):
                    self.log.info('resetting property value %s', key)
                    self._set_property(key, reset=True)
                self.log.info('showing property value   %s', key)
                req = DescribeProperties.from_other(self, Property=[key])
                response = req.main()
                for prop_dict in response.get('properties') or []:
                    if self.args.get('show_defaults'):
                        prop = _build_property(prop_dict.get('name'),
                                               prop_dict.get('defaultValue'),
                                               log=self.log)
                    elif self.args.get('show_descriptions'):
                        # Descriptions are plain text, so force plain
                        # text handling.
                        prop = _build_property(prop_dict.get('name'),
                                               prop_dict.get('description'),
                                               prop_type=_Property,
                                               log=self.log)
                    else:
                        prop = _build_property(prop_dict.get('name'),
                                               prop_dict.get('value'),
                                               log=self.log)
                    if not self.args.get('suppress_all'):
                        prop.print_(
                            suppress_name=self.args.get('suppress_name'),
                            force=self.args.get('all_types'))

    def _set_property(self, prop_name, value=None, reset=None):
        if value == '':
            value = EMPTY
        request = ModifyPropertyValue.from_other(self, Name=prop_name,
                                                 Value=value, Reset=reset)
        response = request.main()
        return response.get('oldValue')

    def _get_all_properties(self, prop_name, prop_type=None):
        # Note that prop_type must currently be able to parse *everything*,
        # so it's probably only safe with the generic _Property class.
        request = DescribeProperties.from_other(self, Property=[prop_name])
        properties = []
        for prop_dict in request.main().get('properties') or []:
            properties.append(_build_property(prop_dict['name'],
                                              prop_dict.get('value'),
                                              prop_type=prop_type,
                                              log=self.log))
        return properties

    def _get_unique_property(self, prop_name, prop_type=None):
        properties = self._get_all_properties(prop_name, prop_type=prop_type)
        if len(properties) < 1:
            raise RuntimeError('no such variable: {0}'.format(prop_name))
        if len(properties) > 1:
            # This is probably where we should implement subtree editing
            # when the service supports it in the future.
            raise RuntimeError('{0} matches more than one variable'
                               .format(prop_name))
        return properties[0]

    def edit_interactively(self, prop):
        formatter = self._get_formatter(prop)
        temp_fd, temp_name = tempfile.mkstemp()
        temp_fd = os.fdopen(temp_fd, 'w')
        try:
            formatter.dump(prop.value, temp_fd)
            editor = os.getenv('VISUAL') or os.getenv('EDITOR') or 'vi'
            while True:
                temp_fd.close()
                subprocess.check_call((editor, temp_name))
                temp_fd = open(temp_name)
                try:
                    # In the event this is a _FallbackProperty, its old
                    # value is very likely to be broken JSON since that
                    # is the only data type eucalyptus uses for
                    # structures.  Since the YAML parser can read the
                    # corrected JSON value we can ignore the fact that
                    # the formatter the user selected with --format will
                    # not match the type of data that he or she entered
                    # in that scenario.  If we later add structures with
                    # new types or we add new --format options then this
                    # will no longer hold, and we had best handle this
                    # case explicitly.
                    new_value = formatter.load(temp_fd)
                except yaml.error.MarkedYAMLError as err:
                    self.log.debug('YAML parse failed', exc_info=True)
                    msg = ('Parsing of new value failed on line {0}: {1}\n{2}'
                           .format(err.problem_mark.line, err.problem,
                                   err.problem_mark.get_snippet()))
                    print >> sys.stderr, msg
                    yesno = raw_input('Edit again? [y/N] ')
                    if not yesno.lower().startswith('y'):
                        raise
                except ValueError as err:
                    # Probably json.loads
                    self.log.debug('New value parse failed', exc_info=True)
                    msg = 'Parsing of new value failed: {0}'.format(
                        err.args[0])
                    print >> sys.stderr, msg
                    yesno = raw_input('Edit again? [y/N] ')
                    if not yesno.lower().startswith('y'):
                        raise
                else:
                    break
        finally:
            temp_fd.close()
            os.remove(temp_name)
        if new_value == prop.value:
            return False  # not changed
        prop.value = new_value
        return True  # changed

    def _get_formatter(self, prop):
        formatter = None
        if self.args.get('format') == 'raw':
            formatter = _Formatter()
        if self.args.get('format') == 'json':
            formatter = _JSONFormatter()
        if self.args.get('format') == 'yaml':
            formatter = _YAMLFormatter()
        if isinstance(prop, _FallbackProperty):
            print >> sys.stderr, 'warning: existing value of', prop.name, \
                'is malformed; outputting it as raw text'
            return _FallbackFormatter(formatter)
        if formatter:
            return formatter
        raise NotImplementedError('formatter "{0}" not implemented'
                                  .format(self.args.get('format')))


def _build_property(prop_name, prop_value, prop_type=None, log=None):
    if prop_type:
        prop = prop_type(prop_name)
    elif PROPERTY_TYPES.get(prop_name) == 'json':
        prop = _JSONProperty(prop_name)
    else:
        prop = _Property(prop_name)
    try:
        prop.loads(prop_value)
    except (ValueError, yaml.error.YAMLError):
        if log:
            log.warn('parsing of variable %s failed; --dump or --edit will '
                     'output it as raw text', prop_name, exc_info=True)
        prop = _FallbackProperty(prop_name, output_formatter=prop.FORMATTER())
        prop.loads(prop_value)
    return prop


class _Formatter(object):
    @staticmethod
    def dumps(value):
        return value

    @staticmethod
    def loads(str_value):
        return str_value

    @classmethod
    def dump(cls, value, fileobj):
        fileobj.write(cls.dumps(value))

    @classmethod
    def load(cls, fileobj):
        return cls.loads(fileobj.read())


class _JSONFormatter(_Formatter):
    @staticmethod
    def dumps(value):
        if not value:
            return ''
        return json.dumps(value, indent=2)

    @staticmethod
    def loads(str_value):
        if not str_value:
            return ''
        if str_value == "{ 'sync': { 'enable':'false' } }":
            # Invalid JSON:  https://eucalyptus.atlassian.net/browse/EUCA-10667
            str_value = str_value.replace("'", '"')
        return json.loads(str_value)


class _YAMLFormatter(_Formatter):
    @staticmethod
    def dumps(value):
        return yaml.safe_dump(value, default_flow_style=False)

    @staticmethod
    def loads(str_value):
        return yaml.safe_load(str_value)


def _FallbackFormatter(input_formatter):
    class AutoFallbackFormatter(_Formatter):
        @staticmethod
        def loads(value):
            return input_formatter.loads(value)

    return AutoFallbackFormatter()


class _Property(object):
    FORMATTER = _Formatter

    def __init__(self, name):
        self.name = name
        self.value = None
        self._formatter = self.FORMATTER()

    def dumps(self):
        if not self.value:
            return ''
        return self._formatter.dumps(self.value)

    def dump(self, fileobj):
        fileobj.write(self.dumps())

    def loads(self, value):
        if value:
            self.value = self._formatter.loads(value)
        else:
            self.value = ''

    def load(self, fileobj):
        self.loads(fileobj.read())

    def print_(self, suppress_name=False, force=False):
        if self.value == {}:
            value = ''
        else:
            value = self.value
        if suppress_name:
            print value.strip()
        else:
            print '{0} = {1}'.format(self.name.strip(), value.strip())


class _JSONProperty(_Property):
    FORMATTER = _JSONFormatter

    def print_(self, suppress_name=False, force=False):
        if force:
            print >> sys.stderr, 'use euctl --dump to view {0}'.format(
                self.name)


class _FallbackProperty(_Property):
    """
    A structure property that uses different formats for input and
    output

    This is used for handling migration from unparseable values to
    parseable values.
    """

    def __init__(self, name, output_formatter=None):
        _Property.__init__(self, name)
        self.__output_formatter = output_formatter or self._formatter

    def dumps(self):
        if not self.value:
            return ''
        return self.__output_formatter.dumps(self.value)

    def loads(self, value):
        _Property.loads(self, value)

    def print_(self, suppress_name=False, force=False):
        if force:
            print >> sys.stderr, 'use euctl --dump to view {0}'.format(
                self.name)
